src/passes/OptimizeInstructions.cpp

```diff
@@ -458,6 +458,91 @@ struct OptimizeInstructions
           return c;
         }
       }
+      {
+        Const* c;
+        Binary* bin;
+        Expression *x, *y;
+        BinaryOp op;
+
+        // fneg(x) * fneg(y)   ==>   x * y
+        // fneg(x) / fneg(y)   ==>   x / y
+        if (matches(
+              curr,
+              binary(&bin, Mul, unary(Neg, any(&x)), unary(Neg, any(&y)))) ||
+            matches(
+              curr,
+              binary(&bin, DivS, unary(Neg, any(&x)), unary(Neg, any(&y))))) {
+          bin->left = x;
+          bin->right = y;
+          return bin;
+        }
+        // fneg(x * C)   ==>   x * -C
+        // fneg(x / C)   ==>   x / -C
+        // fneg(C / x)   ==>   -C / x
+        //
+        // TODO: using `fastMath` is not necessary, but due to fuzzer can't
+        // flexible handle sign of NaNs we use it here.
+        if (fastMath &&
+            (matches(curr, unary(Neg, binary(&bin, Mul, any(&x), fval(&c)))) ||
+             matches(curr, unary(Neg, binary(&bin, DivS, any(&x), fval(&c)))) ||
+             matches(curr,
+                     unary(Neg, binary(&bin, DivS, fval(&c), any(&x)))))) {
+          c->value = c->value.neg();
+          return bin;
+        }
+        // (fneg(x) * y) op C   ==>   (x * y) op -C
+        // (x * fneg(y)) op C   ==>   (x * y) op -C
+        // (fneg(x) / y) op C   ==>   (x / y) op -C
+        // (x / fneg(y)) op C   ==>   (x / y) op -C
+        //
+        // where op = `*` `/` or `+`.
```

Comment:
I don't think `op` should be allowed to be "+" here, as this is not true:
```
(x / fneg(y)) + C   ==>   (x / y) + -C
```
It's only true for "*" and "/".

(There doesn't seem to be a testcase for "+", or the fuzzer would probably have complained.)

Reply:
You're right. Need hide this under `fastMath` flag

Reply:
No, I don't think this is a fastMath issue. That rule is false even with fastMath:
```
(x / fneg(y)) + C   ==>   (x / y) + -C
```
For example, try `x = 1, y = 2, C = 3`, and you get
```
(1 / -2) + 3   !==   1 / 2 + -3
2.5   !==   -2.5
```

Reply:
Ah right! It should be `C - (x / y)` instead. Thanks!